### PR TITLE
dcnm_vrf: Create, Attach and Deploy

### PR DESCRIPTION
**Pull request by nkshrishail**
##### SUMMARY
This PR adds support for a new module for DCNM integration - dcnm_vrf. It covers three options under dcnm_vrf - create, attach and deploy.
Create - Create a single new VRF on an existing fabric. 
Attach - Attach a single VRF to one or more switches on an existing fabric. 
Deploy - Deploy a single VRF to an existing fabric.

##### ISSUE TYPE
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
dcnm

### CODE REVIEW

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (chrisvanheuveln) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py]: I *think* that ansible has a style enforcement that will force us to change from camelCase to snake_case.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (chrisvanheuveln) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py]: I'm running into syntax issues with `networkTemplateConfig` (which is required for the network api). Have you tried populating `vrfTemplateConfig` and using it from a playbook?

edit: not suggesting you need that at this point in development, just asking for my own selfish purposes. :-)

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (chrisvanheuveln) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py]: nit: consider setting the basic path here and then appending to it in the `if` blocks further down; e.g.

```
    path = '/rest/top-down/fabrics/{}/vrf'.format(fabric)
```

```
    if action == 'create':
           ...
            result['response'] = conn.send_request(method, path, json_data)
          ...
    elif action == 'attach':
           ...
            path += '/attachments
            result['response'] = conn.send_request(method, path, json_data)
          ...

```

**Comment by chrisvanheuveln** [comment]
I like your code layout, nice and clean. I suspect the nuances are going to greatly complicate our code eventually but this framework looks pretty solid.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py]: Ok, will change that. Thanks.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py]: It's totally fine, I will give it a try, will have to do it anyway.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py]: Can do this, makes it neat...

**Review by nkshrishail — COMMENTED** [review]
Thanks, thats right, as it will age, it will lose the looks ;)

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (chrisvanheuveln) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py]: Nit: This block feels a little noisy. Below is more concise but it's a matter of taste I guess:

```
    res = result['response']
    if res and isinstance(res, list) and res[0].get('ERROR'):
        vrf_dupe = "VRF ID is already in use" if (action == 'create') else ''
        module.fail_json(msg=[res[0], vrf_dupe])
```

Again, your call. The rest of your code looks good. Merge away. 👍

**Review by chrisvanheuveln — APPROVED** [review]

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py]: This is good, I will update the if conditions on both PRs before merging. Thanks.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:5]: Just adding a reminder that we need to talk about licensing for our software since this will eventually end up in a Cisco owned collection and not part of `ansible/ansible`.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:23]: Update copyright `2020`

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:24]: We need to remove reference to the `Sample Code License` since this will be production code when we release it as part of our collection.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:135]: Found `logit` in the wild! Need to remove this debug code.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:141]: Seems like this method should have a default return value in the event that response is neither a `dict` or `list`?  Probably return `False`?

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:154]: You could still end up with an index error if response is an empty list.  I would prefer the following.

```python
if isinstance(response, list):
    if response and response[0]:
```

Is it possible to end up with a list with elements indicating the vrf object exists?

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:150]: Should you do a `response.get('vrfName')` just to be safe?

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:259]: Is there a way we can query to see if something is attached already and not try attaching again if so?

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (mikewiebe) [inline lib/ansible/modules/network/dcnm/dcnm_vrf.py:264]: Is there a way we can query to see if something is deployed already and not try deploying again if so?

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:23]: Done.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:141]: Yeah, right. Will return False

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:154]: Yeah, missed it in the first round.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:150]: Right.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:259]: Yeah, we do have an API to query the current vrf attachments. The attach VRF API itself does not error if we try the same attachment again. Wondering whats the value add if we put in a check before calling attach API.

File: lib/ansible/modules/network/dcnm/dcnm_vrf.py
Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:264]: There isnt a straight fwd query api available, but, I am yet to exhaust all possible options, I guess there should be way to figure it out. But, again, the deploy API is idempotent like the attach API.

Comment (mikewiebe) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:264]: Ok, so my goal here is to make sure the operations are idempotent.  How were you able to determine that the deploy and attach are already idempotent?

Comment (nkshrishail) [reply lib/ansible/modules/network/dcnm/dcnm_vrf.py:264]: The dcnm attach and deploy APIs are actually idempotent, I didnt have to add additional checks for that. Let me get some sample o/p if that helps.

**Pull request merged**
